tools/python/util/mobile_helpers/usability_checker.py

```diff
@@ -0,0 +1,558 @@
+        else:
+            logger.info(f"{ep_name} is not recommended with this model as there are {self.num_partitions} partitions "
+                        f"covering {pct_nodes_using_ep}% of the nodes in the model. "
+                        "This will most likely result in worse performance than just using the CPU EP.")
+
+
+def check_partitioning(graph: onnx.GraphProto, supported_ops_checker: _SupportedOpsChecker,
+                       require_fixed_input_sizes: bool = False, value_info: dict = None):
+    '''
+    Estimate the partitions the graph will be split into for nodes that is_node_supported_fn returns true for.
+
+    The check on whether a node is supported is purely based on the operator type. Additional limitations
+    (e.g. NNAPI EP only supports 2D Conv) are not checked, so partitions may not be 100% accurate. The limitations
+    for operators in the partitions are printed so the user can manually check.
+    :param graph: Graph to process
+    :param supported_ops_checker: Checker with info on supported ops.
+    :param require_fixed_input_sizes: If True, require that the inputs to a potentially supported node are
+                                       fixed size tensors for it to be considered as supported.
+                                       If True, onnx.shape_inference.infer_shapes should have been run on the model
+                                       to populate the shape information.
+    :param value_info: Map of value name to ValueInfoProto. Required if require_fixed_input_sizes is True to lookup
+                       the shape of a value.
+    :return PartitioningInfo instance with details
+    '''
+
+    if require_fixed_input_sizes and not value_info:
+        raise ValueError("value_info must be provided if require_fixed_input_sizes is True.")
+
+    node_to_producers, node_to_consumers = get_producer_consumer_maps(graph)
+
+    # initializers have fixed sizes.
+    # TODO: when adding subgraph support we also need to match against initializers in ancestor graphs as they are
+    # be accessible from the outer scope (unless shadowed locally)
+    initializers = [i.name for i in graph.initializer]
+
+    def _is_fixed_shape_value(value):
+        if value in value_info:
+            return is_fixed_size_tensor(value_info[value])
+        if value in initializers:
+            return True
+
+        # if something has an unknown shape (e.g. something downstream of a Reshape with dynamic input for the shape)
+        # it won't have an entry in value_info
+        return False
+
+    #
+    # Replicate logic from /onnxruntime/core/providers/partitioning_utils.cc:CreateSupportedPartitionNodeGroups
+    # to roughly estimate number of partitions for nodes that is_node_supported_fn returns true for.
+    #
+    # We keep the structure and variable names as close as possible to the C++ implementation to simplify keeping them
+    # in sync if future updates are needed.
```

Comment:
perhaps add a comment to the C++ implementation so whoever updates it knows about this one too